# Pandas  Part 2 : DataFrame(약어- DF)

* 참고) 사용 준비는 Pandas Part 1 참고!!!

In [1]:
import pandas as pd
import numpy as np

## DF 생성하기 연습

### 자료만 입력 & 컬럼 이름 추가

In [2]:
# DataFrame
d = [0,1,2,3,4]
df = pd.DataFrame(d)
print(df)
print("-----")
ds = pd.Series(d)
print(ds)

   0
0  0
1  1
2  2
3  3
4  4
-----
0    0
1    1
2    2
3    3
4    4
dtype: int64


In [4]:
ds.values

array([0, 1, 2, 3, 4], dtype=int64)

In [7]:
# DataFrame
d = [0,1,2,3,4]
df = pd.DataFrame(d)
print(df)
print("---------")
df.columns = ["column1"]
print(df)

   0
0  0
1  1
2  2
3  3
4  4
---------
   column1
0        0
1        1
2        2
3        3
4        4


### 컬럼 추가하면서 데이터 입력

In [6]:
df["newCol"] = d
print(df)
print("----------------")
df["newCol2"] = 5
print(df)

   0  newCol
0  0       0
1  1       1
2  2       2
3  3       3
4  4       4
----------------
   0  newCol  newCol2
0  0       0        5
1  1       1        5
2  2       2        5
3  3       3        5
4  4       4        5


### 아래의 2가지 문제에 대해서 해결하세요!! -> numpy, pandas에 기본적인 특징을 고민을 하면 쉽게 해결이 가능함!!!! --> "일괄적용"부분!!!

* Q1) df 에서 newCol3 을 기존의 newCol2의 값을 기준으로 만들고, 기존의 newCol2 의 모든 값을 100배 한 값을 newCol3에 생성

* Q2) df에서 newCol2에서 2번째 값을 확인(5)하고, 이 값을 2000으로 변경하기

In [9]:
df[" "]= df["newCol2"]*100
df

0  newCol  newCol2  newCol3
0  0       0        5      500
1  1       1        5      500
2  2       2        5      500
3  3       3        5      500
4  4       4        5      500

In [11]:
df.iat[1,2]  = 2000
df

0  newCol  newCol2  newCol3
0  0       0        5      500
1  1       1     2000      500
2  2       2        5      500
3  3       3        5      500
4  4       4        5      500

In [13]:
df.columns

Index([0, 'newCol', 'newCol2', 'newCol3'], dtype='object')

In [14]:
del df["newCol"]
df

0  newCol2  newCol3
0  0        5      500
1  1     2000      500
2  2        5      500
3  3        5      500
4  4        5      500

In [15]:
df.drop("newCol2", axis=1)

0  newCol3
0  0      500
1  1      500
2  2      500
3  3      500
4  4      500

In [17]:
df.drop(3, axis=0)

0  newCol2  newCol3
0  0        5      500
1  1     2000      500
2  2        5      500
4  4        5      500

In [19]:
df.index

RangeIndex(start=0, stop=5, step=1)

In [20]:
df.columns

Index([0, 'newCol2', 'newCol3'], dtype='object')

In [21]:
df.index.name="count"
df

0  newCol2  newCol3
count                     
0      0        5      500
1      1     2000      500
2      2        5      500
3      3        5      500
4      4        5      500

In [ ]:
변수 이름 = pd.DataFrame([~~~,~~~,~~~~,~~~~,] index=[~~~~,~~~~,~~~~,~~~~~] )

In [22]:
stock_price_DF_withIndex = pd.DataFrame(
    [10000, 10300, 9900, 10500, 11000], 
    index = ["2020-03-01","2020-03-02",
            "2020-03-03","2020-03-04","2020-03-5"])
stock_price_DF_withIndex

0
2020-03-01  10000
2020-03-02  10300
2020-03-03   9900
2020-03-04  10500
2020-03-5   11000

In [24]:
stock_price_DF_withIndex.rename(columns = {0 : "시가"}, inplace=True)
stock_price_DF_withIndex

시가
2020-03-01  10000
2020-03-02  10300
2020-03-03   9900
2020-03-04  10500
2020-03-5   11000

In [26]:
stock_price_DF_withIndex.rename(columns = {"시가2" : "시가"}, inplace=True)
stock_price_DF_withIndex

시가
2020-03-01  10000
2020-03-02  10300
2020-03-03   9900
2020-03-04  10500
2020-03-5   11000

In [27]:
stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

stock_price_DF_withIndex

시가     종가     고가    저가
2020-03-01  10000  11000  13000  3000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000
2020-03-04  10500  13000  18000  8000
2020-03-5   11000  12000  17000  7000

In [ ]:
df이름[컬럼이름] = 내가 던질 값들이 여러개이니....리스트로???
df["종가"] = [11000, 12000, 10000,13000,12000]

In [28]:
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
    index = ["2020-03-01","2020-03-02","2020-03-03","2020-03-04","2020-03-5"])
stock_price_DF_withIndex.rename(columns = {0 : "시가"}, inplace=True)

stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

stock_price_DF_withIndex

시가     종가     고가    저가
2020-03-01  10000  11000  13000  3000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000
2020-03-04  10500  13000  18000  8000
2020-03-5   11000  12000  17000  7000

In [31]:
stock_price_DF_withIndex.iat[0,2]

13000

* Q) 2020년 3월 1일의 시가 / 종가를 출력해보세요...
* Q) 2020년 3월 1일 에서 2020년 3월 3일의 시가/종가/고가/저가를 출력해 보세요!!!

In [ ]:
stock_price_DF_withIndex.loc[가로, 세로]

In [32]:
stock_price_DF_withIndex.loc["2020-03-01", ["시가", "종가"]]

시가    10000
종가    11000
Name: 2020-03-01, dtype: int64

In [33]:
stock_price_DF_withIndex.loc["2020-03-01":"2020-03-03", "시가":"저가"]

시가     종가     고가    저가
2020-03-01  10000  11000  13000  3000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000

In [34]:
stock_price_DF_withIndex.loc["2020-03-01":"2020-03-03", :]

시가     종가     고가    저가
2020-03-01  10000  11000  13000  3000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000

In [47]:
stock_price_DF_withIndex.values

array([[10000, 11000, 13000,  3000],
       [10300, 12000, 15000,  5000],
       [ 9900, 10000, 15000,  5000],
       [10500, 13000, 18000,  8000],
       [11000, 12000, 17000,  7000]], dtype=int64)

In [42]:
stock_price_DF_withIndex.shape[0]

5

In [43]:
stock_price_DF_withIndex.shape[1]

4

In [37]:
stock_price_DF_withIndex[["시가","종가"]]

시가     종가
2020-03-01  10000  11000
2020-03-02  10300  12000
2020-03-03   9900  10000
2020-03-04  10500  13000
2020-03-5   11000  12000

In [48]:
obj = pd.Series([4.5, 7.2, -5.3], index=["c","b","a"])
obj

c    4.5
b    7.2
a   -5.3
dtype: float64

In [54]:
obj.reindex(["a","d","e","c","b"]).interpolate()

a   -5.300000
d   -2.033333
e    1.233333
c    4.500000
b    7.200000
dtype: float64

In [51]:
obj2 = obj.reindex(["a","b","c","d","e"],fill_value=99)
obj2

a    -5.3
b     7.2
c     4.5
d    99.0
e    99.0
dtype: float64

In [61]:
stock_price_DF_withIndex.sort_index(axis=1, ascending=False).sort_index(axis=0, ascending=False)

종가    저가     시가     고가
2020-03-5   12000  7000  11000  17000
2020-03-04  13000  8000  10500  18000
2020-03-03  10000  5000   9900  15000
2020-03-02  12000  5000  10300  15000
2020-03-01  11000  3000  10000  13000

In [67]:
stock_price_DF_withIndex.sort_values(by="고가", ascending=False)

시가     종가     고가    저가
2020-03-04  10500  13000  18000  8000
2020-03-5   11000  12000  17000  7000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000
2020-03-01  10000  11000  13000  3000

In [66]:
stock_price_DF_withIndex.sort_values(by=["고가","종가"], ascending=False)

시가     종가     고가    저가
2020-03-04  10500  13000  18000  8000
2020-03-5   11000  12000  17000  7000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000
2020-03-01  10000  11000  13000  3000

In [68]:
stock_price_DF_withIndex

시가     종가     고가    저가
2020-03-01  10000  11000  13000  3000
2020-03-02  10300  12000  15000  5000
2020-03-03   9900  10000  15000  5000
2020-03-04  10500  13000  18000  8000
2020-03-5   11000  12000  17000  7000

In [73]:
stock_price_DF_withIndex.applymap(lambda x: x+100)

시가     종가     고가    저가
2020-03-01  10100  11100  13100  3100
2020-03-02  10400  12100  15100  5100
2020-03-03  10000  10100  15100  5100
2020-03-04  10600  13100  18100  8100
2020-03-5   11100  12100  17100  7100

In [72]:
stock_price_DF_withIndex.apply(lambda x: x.max() - x.min(), axis=1)

2020-03-01    10000
2020-03-02    10000
2020-03-03    10000
2020-03-04    10000
2020-03-5     10000
dtype: int64

In [86]:
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
    index = ["2020-03-01","2020-03-02","2020-03-03","2020-03-04","2020-03-5"])
stock_price_DF_withIndex.rename(columns = {0 : "시가"}, inplace=True)

stock_price_DF_withIndex["종가"] = [350,234,234000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,122000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

stock_price_DF_withIndex

시가      종가      고가    저가
2020-03-01  10000     350   13000  3000
2020-03-02  10300     234   15000  5000
2020-03-03   9900  234000   15000  5000
2020-03-04  10500   13000  122000  8000
2020-03-5   11000   12000   17000  7000

In [87]:
stock_price_DF_withIndex["Gap"] = stock_price_DF_withIndex.apply(lambda x:x.max()-x.min(), axis=1)

stock_price_DF_withIndex

시가      종가      고가    저가     Gap
2020-03-01  10000     350   13000  3000   12650
2020-03-02  10300     234   15000  5000   14766
2020-03-03   9900  234000   15000  5000  229000
2020-03-04  10500   13000  122000  8000  114000
2020-03-5   11000   12000   17000  7000   10000

In [78]:
len(stock_price_DF_withIndex)

5

In [80]:
stock_price_DF_withIndex.shape[0]

5

In [81]:
stock_price_DF_withIndex["Gap_forloop"] =0
stock_price_DF_withIndex

시가      종가      고가    저가     Gap  Gap_forloop
2020-03-01  10000     350   13000  3000   12650            0
2020-03-02  10300     234   15000  5000   14766            0
2020-03-03   9900  234000   15000  5000  229000            0
2020-03-04  10500   13000  122000  8000  114000            0
2020-03-5   11000   12000   17000  7000   10000            0

In [ ]:
df[새로운 컬럼 이름] = 0
stock_price_DF_withIndex["gap_forloop"] = 0

for i in range(0, len(데이터프레임)):
for i in range(0, len(stock_price_DF_withIndex)):
    
    temp:i 번째 줄에서...앞의 4개의 컬럼의 값을 가지고 와야 함
    (가로 i번째 줄, 세로는 시가/종가/저가/고가) -->  iloc[i, ~~~]
    temp = stock_price_DF_withIndex.iloc[i , 0:4]
    
    위의 4개의 값 중(temp)에서 최고값 찾기 : max() --> tempMax 변수에..
    tempMax = temp.max()
    위의 4개의 값 중(temp)에서 최저값 찾기 : min() --> tempmin 변수에..
    tempmin = temp.min()
    가로 i번째 줄의 4개 값 중에서 최고값 - 최저값 : tempMax - tempmin
    
    가로 i번째, 세로 gap_forloop 의 셀 위치에 tempMax - tempmin 값 대입!!
    stock_price_DF_withIndex.iat[i, 5] = tempMax - tempmin



In [88]:
stock_price_DF_withIndex["gap_forloop"] = 0
for i in range(0, len(stock_price_DF_withIndex)):
    temp = stock_price_DF_withIndex.iloc[i , 0:4]
    stock_price_DF_withIndex.iat[i, 5] = temp.max() - temp.min()
stock_price_DF_withIndex

시가      종가      고가    저가     Gap  gap_forloop
2020-03-01  10000     350   13000  3000   12650        12650
2020-03-02  10300     234   15000  5000   14766        14766
2020-03-03   9900  234000   15000  5000  229000       229000
2020-03-04  10500   13000  122000  8000  114000       114000
2020-03-5   11000   12000   17000  7000   10000        10000

In [ ]:
stock_price_DF_withIndex["gap_forloop"] = 0
for i in range(0, len(stock_price_DF_withIndex)):
    temp = stock_price_DF_withIndex.iloc[i , 0:4]
    stock_price_DF_withIndex.iat[i, stock_price_DF_withIndex.columns.get_loc("gap_forloop")] = temp.max() - temp.min()
stock_price_DF_withIndex

In [93]:
labels = ["시가", "고가","종가"]

labels_idx = [stock_price_DF_withIndex.columns.get_loc(label) for label in labels] 
labels_idx

[0, 2, 1]

In [ ]:
stock_price_DF_withIndex["gap_forloop"] = 0

labels = ["gap_forloop"]
labels_idx = [stock_price_DF_withIndex.columns.get_loc(i) for i in labels ]
cols = ["시가", "종가", "저가", "고가"]
cols_idx = [stock_price_DF_withIndex.columns.get_loc(i) for i in cols]

for i in range(0, len(stock_price_DF_withIndex)):
    temp = stock_price_DF_withIndex.iloc[i ,cols_idx ]
    stock_price_DF_withIndex.iat[i,labels_idx[0]] = temp.max() - temp.min()
stock_price_DF_withIndex

In [ ]:
for i in stock_price_DF_withIndex.index:
    temp = stock_price_DF_withIndex.loc[i, ["시가","종가","고가","저가"]]
    gap = temp.max() - temp.min()
    stock_price_DF_withIndex.at[i, "gap_forloop2"] = gap
stock_price_DF_withIndex

In [4]:
# Q1)
df["newCol3"] = df["newCol2"]*100
df

column1  newCol  newCol2  newCol3
0        0       0        5      500
1        1       1        5      500
2        2       2        5      500
3        3       3        5      500
4        4       4        5      500

* pandas update로 인해서 ix는 사용하면 경고가 나타남!!!

In [5]:
# Q2) 
print(df.ix[1,2])
print("------")
df.ix[1,2] = 2000
print("------")


5
------
------


/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:2: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  after removing the cwd from sys.path.


In [14]:
# Q2) 
print(df.iloc[1,2])
print("------")
df.iloc[1,2] = 2000
print("------")
print(df)

5
------
------
   column1  newCol  newCol2  newCol3
0        0       0        5      500
1        1       1     2000      500
2        2       2        5      500
3        3       3        5      500
4        4       4        5      500


### 컬럼 제거하기 : 여러가지 방법 존재

#### 방법 1 ) del 함수 이용

In [7]:
# 컬럼 제거 방법1 : del function
print(df)
del df["column1"]
print(df)

   column1  newCol  newCol2  newCol3
0        0       0        5      500
1        1       1     2000      500
2        2       2        5      500
3        3       3        5      500
4        4       4        5      500
   newCol  newCol2  newCol3
0       0        5      500
1       1     2000      500
2       2        5      500
3       3        5      500
4       4        5      500


#### 필요한 컬럼 선택으로 간접 제거

In [8]:
# 컬럼제거 방법2 : multi-column select
print(df.columns.values)
newColList = [
    'newCol2', 
    'newCol3'
]
df[newColList]

['newCol' 'newCol2' 'newCol3']


newCol2  newCol3
0        5      500
1     2000      500
2        5      500
3        5      500
4        5      500

#### Q) df = df["newColList] 와 아래의 df[newColList]의 차이점은???? 위에서 df 를 직접 확인해보면서 이해하기!!!

In [9]:

# 주의!!)
# df = df["newColList] 와 아래의 df[newColList]의 차이점은????
print(df[newColList])


#df = df[["newCol2","newCol3"]]

   newCol2  newCol3
0        5      500
1     2000      500
2        5      500
3        5      500
4        5      500


#### 방법3 : drop 사용

In [10]:
# 컬럼제거 방법3 : drop 메소드 사용
print (df.drop("newCol", axis=1))
print (df.drop(3, axis=0))


   newCol2  newCol3
0        5      500
1     2000      500
2        5      500
3        5      500
4        5      500
   newCol  newCol2  newCol3
0       0        5      500
1       1     2000      500
2       2        5      500
4       4        5      500


### DF에서 index 이름 만들기! --> Series 와 크게 다르지 않음!!!

In [11]:
# index 이름
df.index.name="Count"
df

newCol  newCol2  newCol3
Count                          
0           0        5      500
1           1     2000      500
2           2        5      500
3           3        5      500
4           4        5      500

### 접근하는 법!!!

* DF에서 Access!!
* 단, 그 전에 데이터를 생성하고, 변경하는 과정을 수행할 것 임!!!

#### 우선 다시 할 자료 만들기

In [16]:
# DataFrame에서 검색
#01-) 자료 만들기
stock_price_DF_withIndex = pd.DataFrame(
    [10000, 10300, 9900, 10500, 11000], 
    index = ["2016-03-01","2016-03-02",
                                                 "2016-03-03","2016-03-04","2016-03-5"])
stock_price_DF_withIndex

0
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000

#### 컬럼 이름 변경

In [17]:
# 컬럼 이름 변경 방법1 --> 변경 되나??? 왜 안 될까????
stock_price_DF_withIndex.rename(columns={"0":"시가"}, inplace = True)
stock_price_DF_withIndex

0
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000

In [18]:
# 컬럼 이름 변경 방법1 - 수정!!!
stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)
stock_price_DF_withIndex

시가
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000

In [15]:
# 컬럼 이름 변경 방법2 
stock_price_DF_withIndex.columns.values[0] = "시가2"
stock_price_DF_withIndex

시가2
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000

#### Q) 위에서는 되는데, 아래는 위의 사항을 동일하게 하였는데 왜 안되는 것인가???

In [16]:
stock_price_DF_withIndex = pd.DataFrame( [10000, 10300, 9900, 10500, 11000], index = ["2016-03-01","2016-03-02","2016-03-03","2016-03-04","2016-03-5"])
stock_price_DF_withIndex
stock_price_DF_withIndex.columns.values[0] = u"시가2"
stock_price_DF_withIndex

ValueError: invalid literal for int() with base 10: '시가2'

In [17]:
print (stock_price_DF_withIndex)
print ("---------")
print (stock_price_DF_withIndex.columns)
print ("---------")
print (stock_price_DF_withIndex.columns.values)
print ("---------")
print (stock_price_DF_withIndex.columns.values[0])

# 숫자로는 변경이 가능하지만, 문자열로는 불가능!!! : RangeIndex 이므로 숫자이니!!!
# stock_price_DF_withIndex.columns.values[0] = 1 은 가능
# stock_price_DF_withIndex.columns.values[0] = "1" 은 에러!!!!

                0
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000
---------
RangeIndex(start=0, stop=1, step=1)
---------
[0]
---------
0


#### 위의 내용을 바탕으로 아래와 같은 데이터를 만들어보자 :  pandas DataFrame으로!!!

* 위의 사항들을 확인하였으며, 정상적인 메소드를 이용해서 위의 사항을 다시 해보자.

* 위의 자료를 활용하여, "시가"를 가진 DF 에서 "종가", "고가","저가"의 컬럼을 추가하여 생성하여 보자!!!



*  	시가 	종가 	고가 	저가
* 2016-03-01 	10000 	11000 	13000 	3000
* 2016-03-02 	10300 	12000 	15000 	5000
* 2016-03-03 	9900 	10000 	15000 	5000
* 2016-03-04 	10500 	13000 	18000 	8000
* 2016-03-5 	11000 	12000 	17000 	7000

In [2]:
# 위의 사항 이용 방법)
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
                                        index = ["2016-03-01","2016-03-02","2016-03-03","2016-03-04","2016-03-5"])
stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)

stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

stock_price_DF_withIndex

시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000

#### 참고) rename이라는 정상적인 메소드를 활용하여 컬럼 이름 변경하기!!!!

In [19]:
# rename 활용!!!
stock_price_DF_withIndex.rename(columns={"시가":"시가2", "종가":"종가2", "고가":"고가2"}, inplace = True)
stock_price_DF_withIndex

시가2    종가2    고가2    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000

In [20]:
# rename 활용하여 다시 원상 복귀!!!
stock_price_DF_withIndex.rename(columns={"시가2":"시가", "종가2":"종가", "고가2":"고가"}, inplace = True)
stock_price_DF_withIndex

시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000

#### 아래의 문제들을 해결해보세요!!!!

* Q) print stock_Price_DF_withIndex.iat[0] 은 왜 에러가 발생할까? 어떻게 수정하면 되는가?

* Q)앞의 데이터에서 2016년 3월 1일의 시가/종가는?

* Q)앞의 데이터에서 2016년 3월 1일 ~ 2016년 3월 3일 사이의 시가/종가/고가/저가는?

In [26]:
print (stock_price_DF_withIndex.loc["2016-03-01", ["시가","종가"]])

시가    10000
종가    11000
Name: 2016-03-01, dtype: int64


In [28]:
print (stock_price_DF_withIndex.loc["2016-03-01":"2016-03-03", "시가":"저가"])

               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000


In [22]:
print (stock_price_DF_withIndex.iat[0,0])

10000


In [21]:
print (stock_price_DF_withIndex.loc["2016-03-01",["시가","종가"]])

시가    10000
종가    11000
Name: 2016-03-01, dtype: int64


In [24]:
print (stock_price_DF_withIndex.loc["2016-03-01":"2016-03-03","시가":"저가"])

               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000


In [21]:
#Access???
#print stock_price_DF_withIndex.iat[0]

print (stock_price_DF_withIndex.iat[0,0])
print (stock_price_DF_withIndex.at["2016-03-01","시가"])
#print stock_price_DF_withIndex.iat[0,0:3]

print (stock_price_DF_withIndex.loc["2016-03-01","시가"])
print (stock_price_DF_withIndex.iloc[0,0])
print (stock_price_DF_withIndex.iloc[0,0:3])

# (((주의!!!!!!)))최신 업데이트에서는 경고문구 나타남!!!!!!
print (stock_price_DF_withIndex.ix[0,0])
print (stock_price_DF_withIndex.ix[0:3,0:3])
print (stock_price_DF_withIndex.ix["2016-03-01":"2016-03-03",0:3])
print (stock_price_DF_withIndex.ix["2016-03-01"])

10000
10000
10000
10000
시가    10000
종가    11000
고가    13000
Name: 2016-03-01, dtype: int64
10000
               시가     종가     고가
2016-03-01  10000  11000  13000
2016-03-02  10300  12000  15000
2016-03-03   9900  10000  15000
               시가     종가     고가
2016-03-01  10000  11000  13000
2016-03-02  10300  12000  15000
2016-03-03   9900  10000  15000
시가    10000
종가    11000
고가    13000
저가     3000
Name: 2016-03-01, dtype: int64


/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:13: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  del sys.path[0]
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:14: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:15: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  from ipykernel import kernelapp as app
/Users/minkyuk

In [22]:
# Search
print (stock_price_DF_withIndex.loc["2016-03-01"])
print (stock_price_DF_withIndex.loc["2016-03-01"]["시가"])
print (stock_price_DF_withIndex.loc["2016-03-01"]["고가"])
print ("------------")
print (stock_price_DF_withIndex.iloc[0])
print (stock_price_DF_withIndex.iloc[0]["시가"])
print (stock_price_DF_withIndex.iloc[0]["고가"])
print ("------------")


시가    10000
종가    11000
고가    13000
저가     3000
Name: 2016-03-01, dtype: int64
10000
13000
------------
시가    10000
종가    11000
고가    13000
저가     3000
Name: 2016-03-01, dtype: int64
10000
13000
------------


#### DF 의 컬럼에 접근해보자!!! 앞에서는 index(정수, 내가 만든 것)을 이용해서 접근을 하였으니..

In [23]:
# 컬럼 접근 -01)
stock_price_DF_withIndex["시가"]

2016-03-01    10000
2016-03-02    10300
2016-03-03     9900
2016-03-04    10500
2016-03-5     11000
Name: 시가, dtype: int64

In [24]:
# 컬럼 접근 -02)
stock_price_DF_withIndex[["시가","종가"]] 

시가     종가
2016-03-01  10000  11000
2016-03-02  10300  12000
2016-03-03   9900  10000
2016-03-04  10500  13000
2016-03-5   11000  12000

In [25]:
# 컬럼 접근 -03) 최신 버전에서는 경고 나타남!!!!!!!!
# 주의!!!!!) 
stock_price_DF_withIndex.ix[:,["시가","종가"]]

/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  This is separate from the ipykernel package so we can avoid doing imports until


시가     종가
2016-03-01  10000  11000
2016-03-02  10300  12000
2016-03-03   9900  10000
2016-03-04  10500  13000
2016-03-5   11000  12000

#### 참고) 다음 아래 두 사항을 비교해볼 것!!! 겉으로는 동일하지만, 나중에 해보면 좀 다른 부분이 존재함!

In [26]:
test = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
                                        index = ["2016-03-01","2016-03-02",
                                                 "2016-03-03","2016-03-04","2016-03-5"])
test

0
2016-03-01  10000
2016-03-02  10300
2016-03-03   9900
2016-03-04  10500
2016-03-5   11000

In [27]:
test = pd.DataFrame({0:[10000, 10300, 9900, 10500, 11000],1:[1,2,3,4,5]}, 
                                        index = ["2016-03-01","2016-03-02",
                                                 "2016-03-03","2016-03-04","2016-03-5"])
test

0  1
2016-03-01  10000  1
2016-03-02  10300  2
2016-03-03   9900  3
2016-03-04  10500  4
2016-03-5   11000  5

In [28]:
test[0]

2016-03-01    10000
2016-03-02    10300
2016-03-03     9900
2016-03-04    10500
2016-03-5     11000
Name: 0, dtype: int64

In [29]:
test[0:1]

0  1
2016-03-01  10000  1

In [31]:
print(stock_price_DF_withIndex.columns)
stock_price_DF_withIndex[0]

Index(['시가', '종가', '고가', '저가'], dtype='object')


KeyError: 0

In [32]:
# 비교!!! 주의!!! 지금은 경고 나타남!!!!!!!!
print (stock_price_DF_withIndex[0:1])
print ("----------------")
print (stock_price_DF_withIndex.ix[0:1])
print ("----------------")
print (stock_price_DF_withIndex.ix[0:1,:])

               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
----------------
               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
----------------
               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000


/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  after removing the cwd from sys.path.
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:6: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  


#### 기타 내용: .T, .values, .shape, .size, .ndim etc..

In [33]:
#  기타 내용 -01
stock_price_DF_withIndex.T

2016-03-01  2016-03-02  2016-03-03  2016-03-04  2016-03-5
시가       10000       10300        9900       10500      11000
종가       11000       12000       10000       13000      12000
고가       13000       15000       15000       18000      17000
저가        3000        5000        5000        8000       7000

In [34]:
#  기타 내용 -02
stock_price_DF_withIndex.values

array([[10000, 11000, 13000,  3000],
       [10300, 12000, 15000,  5000],
       [ 9900, 10000, 15000,  5000],
       [10500, 13000, 18000,  8000],
       [11000, 12000, 17000,  7000]])

In [35]:
#  기타 내용 -03
stock_price_DF_withIndex.shape

(5, 4)

In [36]:
stock_price_DF_withIndex.shape[0]

5

In [37]:
stock_price_DF_withIndex.shape[1]

4

In [38]:
#  기타 내용 -04
stock_price_DF_withIndex.size

20

In [39]:
#  기타 내용 -04
stock_price_DF_withIndex.ndim

2

### DF  : reindex !!!

In [29]:
### ReIndex
obj = pd.Series([4.5,7.2,-5.3], index=["c","b","a"])
obj

c    4.5
b    7.2
a   -5.3
dtype: float64

In [30]:
obj2 = obj.reindex(["a","b","c","d","e"])
obj2

a   -5.3
b    7.2
c    4.5
d    NaN
e    NaN
dtype: float64

In [31]:
obj.reindex(["a","b"])

a   -5.3
b    7.2
dtype: float64

In [32]:
obj

c    4.5
b    7.2
a   -5.3
dtype: float64

In [43]:
obj.reindex(["a","b","c","d","e"], fill_value=0)

a   -5.3
b    7.2
c    4.5
d    0.0
e    0.0
dtype: float64

#### 참고) method : {None, ‘backfill’/’bfill’, ‘pad’/’ffill’, ‘nearest’}, optional

    method to use for filling holes in reindexed DataFrame. Please note: this is only applicable to DataFrames/Series with a monotonically increasing/decreasing index.
    이런 이유로 index가 순서상 문제가 있으면, method 의 방법이 제대로 안 될 수 있음!!!
    
예) obj = Series([4.5,7.2,-5.3], index=["c","b","a"])
print obj
print obj.reindex(["a","d","e","c","b"])
print obj.reindex(["a","d","e","c","b"], method='ffill')
print obj.reindex(["a","d","e","c","b"], method='bfill')


In [44]:
obj3 = pd.Series(['blue', 'purple', 'yellow'], index=[0, 2, 4])
print(obj3)
print("--------------")
print(obj3.reindex(range(6), method='ffill'))

0      blue
2    purple
4    yellow
dtype: object
--------------
0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
dtype: object


In [45]:
print(obj3.reindex([0,1,2,3,4,5,6,7,8]))
print("--------------")
print(obj3.reindex([0,1,2,3,4,5,6,7,8],method='ffill'))
print("--------------")
print(obj3.reindex([0,1,2,3,4,5,6,7,8],method='bfill'))

0      blue
1       NaN
2    purple
3       NaN
4    yellow
5       NaN
6       NaN
7       NaN
8       NaN
dtype: object
--------------
0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
6    yellow
7    yellow
8    yellow
dtype: object
--------------
0      blue
1    purple
2    purple
3    yellow
4    yellow
5       NaN
6       NaN
7       NaN
8       NaN
dtype: object


### 실질적으로 "보간법"으로 처리하는 경우도 많음!!

In [46]:
# 보간법
obj = pd.Series([4.5,7.2,-5.3], index=["c","b","a"])
print(obj)
print("--------------")
print(obj.reindex(["a","d","e","c","b"]))
print("--------------")
print(obj.reindex(["a","d","e","c","b"]).interpolate())

c    4.5
b    7.2
a   -5.3
dtype: float64
--------------
a   -5.3
d    NaN
e    NaN
c    4.5
b    7.2
dtype: float64
--------------
a   -5.300000
d   -2.033333
e    1.233333
c    4.500000
b    7.200000
dtype: float64


#### Q) 앞의 데이터를 가지고, Series가 아닌 DataFrame으로 만들고 나서, reindex는 앞에서와 같이 “a”,”d”,”e”,”c”,”b”로 하고, reindex한 결과 보이기 & “interPo”라는 새로운 칼럼을 만들어서 interpolate를 한 값으로 채우기



* 아래와 같은 DF를 만들어 보세요!!
*      0   interPo
* a -5.3 -5.300000
* d  NaN -2.033333
* e  NaN  1.233333
* c  4.5  4.500000
* b  7.2  7.200000


In [47]:
# 보간법 실습
obj = pd.DataFrame([4.5,7.2,-5.3], index=["c","b","a"])
obj = obj.reindex(["a","d","e","c","b"])
print(obj)
print("--------------")
# pnadas update로 인해서 ix 경고로 수정 : 2018.08
#obj["interPo"] = obj.ix[:,0].interpolate()
obj["interPo"] = obj.iloc[:,0].interpolate()
print(obj)

     0
a -5.3
d  NaN
e  NaN
c  4.5
b  7.2
--------------
     0   interPo
a -5.3 -5.300000
d  NaN -2.033333
e  NaN  1.233333
c  4.5  4.500000
b  7.2  7.200000


In [48]:
# 참고) 그냥 보간법만 돌리면 만사 OK??
import numpy as np
from numpy import nan as NA
df = pd.DataFrame(np.random.rand(7,3))
df

0         1         2
0  0.022382  0.789548  0.993454
1  0.730255  0.469351  0.423037
2  0.008122  0.802727  0.134072
3  0.304750  0.775250  0.207822
4  0.502088  0.368350  0.602600
5  0.618860  0.148258  0.779752
6  0.267541  0.014843  0.483034

In [49]:
# pandas update관련으로 ix수정 : 2018.08
df.iloc[0,1] = NA
df.iloc[1,1] = NA
df.iloc[2,1] = NA
df.iloc[3,1] = NA
df.iloc[4,1] = NA

df.iloc[0,2] = NA
df.iloc[1,2] = NA
df.iloc[2,2] = NA
df

0         1         2
0  0.022382       NaN       NaN
1  0.730255       NaN       NaN
2  0.008122       NaN       NaN
3  0.304750       NaN  0.207822
4  0.502088       NaN  0.602600
5  0.618860  0.148258  0.779752
6  0.267541  0.014843  0.483034

In [50]:
# 결과가 맞나????
df.interpolate()

0         1         2
0  0.022382       NaN       NaN
1  0.730255       NaN       NaN
2  0.008122       NaN       NaN
3  0.304750       NaN  0.207822
4  0.502088       NaN  0.602600
5  0.618860  0.148258  0.779752
6  0.267541  0.014843  0.483034

In [51]:
# 참고) 이런 경우에는 이상적으로 보간법이 잘 동작
import numpy as np
from numpy import nan as NA
df = pd.DataFrame(np.random.rand(7,3))
df

0         1         2
0  0.691602  0.670400  0.494303
1  0.512938  0.236714  0.733266
2  0.444326  0.838333  0.250761
3  0.483143  0.378696  0.769861
4  0.627272  0.199859  0.728673
5  0.760880  0.951057  0.676104
6  0.567814  0.788994  0.358032

In [52]:
# update : 2018.8 --> pandas change ix method
# df.ix[1,1] = NA
# df.ix[3,1] = NA
# df.ix[2,2] = NA
# df.ix[5,2] = NA
df.iloc[1,1] = NA
df.iloc[3,1] = NA
df.iloc[2,2] = NA
df.iloc[5,2] = NA

df

0         1         2
0  0.691602  0.670400  0.494303
1  0.512938       NaN  0.733266
2  0.444326  0.838333       NaN
3  0.483143       NaN  0.769861
4  0.627272  0.199859  0.728673
5  0.760880  0.951057       NaN
6  0.567814  0.788994  0.358032

In [53]:
df.interpolate()

0         1         2
0  0.691602  0.670400  0.494303
1  0.512938  0.754367  0.733266
2  0.444326  0.838333  0.751563
3  0.483143  0.519096  0.769861
4  0.627272  0.199859  0.728673
5  0.760880  0.951057  0.543352
6  0.567814  0.788994  0.358032

### 정렬/ 순위

In [ ]:
# 앞에 데이터가 있으면 다시 하지 않아도 됨!

# stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
#                                         index = ["2016-03-03","2016-03-01","2016-03-02","2016-03-5 ","2016-03-04"])
# stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)

# stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
# stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
# stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

# print stock_price_DF_withIndex

In [8]:
print(stock_price_DF_withIndex)

               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000


#### 아래 옵션들을 보면서 어떻게 동작하는지 이해할 것!: 단순, 가로 이름 기준/세로 이름 기준 ( 주의! "값"이 아님!!!)

* 기준이 index도 있고, value도 있음!!!!

In [14]:
print(stock_price_DF_withIndex)
print("----------------")
print(stock_price_DF_withIndex.sort_index(axis=1))
# print stock_price_DF_withIndex.sort_index(axis=0)
# print stock_price_DF_withIndex.sort_index(axis=1)
# print stock_price_DF_withIndex.sort_index(axis=0, ascending = True)
# print stock_price_DF_withIndex.sort_index(axis=0, ascending = False)

# print stock_price_DF_withIndex.sort_index(axis=1, ascending = True)
# print stock_price_DF_withIndex.sort_index(axis=1, ascending = False)

               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000
----------------
               고가     시가    저가     종가
2016-03-01  13000  10000  3000  11000
2016-03-02  15000  10300  5000  12000
2016-03-03  15000   9900  5000  10000
2016-03-04  18000  10500  8000  13000
2016-03-5   17000  11000  7000  12000


In [55]:
stock_price_DF_withIndex.sort_index

<bound method DataFrame.sort_index of                시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000>

In [56]:
stock_price_DF_withIndex.sort_index(axis=0)

시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000

In [57]:
stock_price_DF_withIndex.sort_index(axis=1)

고가     시가    저가     종가
2016-03-01  13000  10000  3000  11000
2016-03-02  15000  10300  5000  12000
2016-03-03  15000   9900  5000  10000
2016-03-04  18000  10500  8000  13000
2016-03-5   17000  11000  7000  12000

In [58]:
stock_price_DF_withIndex.sort_index(axis=0, ascending = True)

시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000

In [59]:
stock_price_DF_withIndex.sort_index(axis=0, ascending = False)

시가     종가     고가    저가
2016-03-5   11000  12000  17000  7000
2016-03-04  10500  13000  18000  8000
2016-03-03   9900  10000  15000  5000
2016-03-02  10300  12000  15000  5000
2016-03-01  10000  11000  13000  3000

In [60]:
stock_price_DF_withIndex.sort_index(axis=1, ascending = True)

고가     시가    저가     종가
2016-03-01  13000  10000  3000  11000
2016-03-02  15000  10300  5000  12000
2016-03-03  15000   9900  5000  10000
2016-03-04  18000  10500  8000  13000
2016-03-5   17000  11000  7000  12000

In [15]:
stock_price_DF_withIndex.sort_index(axis=1, ascending = False)

종가    저가     시가     고가
2016-03-01  11000  3000  10000  13000
2016-03-02  12000  5000  10300  15000
2016-03-03  10000  5000   9900  15000
2016-03-04  13000  8000  10500  18000
2016-03-5   12000  7000  11000  17000

In [16]:
stock_price_DF_withIndex.sort_index(axis=1, ascending = False).sort_index(axis=0, ascending=False)

종가    저가     시가     고가
2016-03-5   12000  7000  11000  17000
2016-03-04  13000  8000  10500  18000
2016-03-03  10000  5000   9900  15000
2016-03-02  12000  5000  10300  15000
2016-03-01  11000  3000  10000  13000

In [62]:
# 양쪽 모두....
stock_price_DF_withIndex.sort_index(axis=0).sort_index(axis=1)

고가     시가    저가     종가
2016-03-01  13000  10000  3000  11000
2016-03-02  15000  10300  5000  12000
2016-03-03  15000   9900  5000  10000
2016-03-04  18000  10500  8000  13000
2016-03-5   17000  11000  7000  12000

#### 정렬 할 떄에, 가로/세로 이름이 아니라 "값"을 중심으로 정렬하고 싶을떄??

In [63]:
#정렬 순위 : sort_value
print (stock_price_DF_withIndex.sort_values(by = "고가"))
print (stock_price_DF_withIndex.sort_values(by = "고가", ascending=False))
print (stock_price_DF_withIndex.sort_values(by = "고가", ascending=True))

# 여러 컬럼에 적용?
print (stock_price_DF_withIndex.sort_values(by = ["고가","종가"], ascending=True))


               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-5   11000  12000  17000  7000
2016-03-04  10500  13000  18000  8000
               시가     종가     고가    저가
2016-03-04  10500  13000  18000  8000
2016-03-5   11000  12000  17000  7000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-01  10000  11000  13000  3000
               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-02  10300  12000  15000  5000
2016-03-03   9900  10000  15000  5000
2016-03-5   11000  12000  17000  7000
2016-03-04  10500  13000  18000  8000
               시가     종가     고가    저가
2016-03-01  10000  11000  13000  3000
2016-03-03   9900  10000  15000  5000
2016-03-02  10300  12000  15000  5000
2016-03-5   11000  12000  17000  7000
2016-03-04  10500  13000  18000  8000


#### 참고) 정렬 : 값을 기준으로 rank 를 정해서 할 경우 - 무엇을 기준으로 할 것인지는 method 활용

In [64]:
#정렬 순위 : rank
stock_price_DF_withIndex["rank_end"] = stock_price_DF_withIndex["종가"].rank()
stock_price_DF_withIndex


시가     종가     고가    저가  rank_end
2016-03-01  10000  11000  13000  3000       2.0
2016-03-02  10300  12000  15000  5000       3.5
2016-03-03   9900  10000  15000  5000       1.0
2016-03-04  10500  13000  18000  8000       5.0
2016-03-5   11000  12000  17000  7000       3.5

In [65]:
stock_price_DF_withIndex["rank_end"] = stock_price_DF_withIndex["종가"].rank(method="max")
stock_price_DF_withIndex


시가     종가     고가    저가  rank_end
2016-03-01  10000  11000  13000  3000       2.0
2016-03-02  10300  12000  15000  5000       4.0
2016-03-03   9900  10000  15000  5000       1.0
2016-03-04  10500  13000  18000  8000       5.0
2016-03-5   11000  12000  17000  7000       4.0

In [66]:
stock_price_DF_withIndex["rank_end"] = stock_price_DF_withIndex["종가"].rank(ascending=False,method="max")
stock_price_DF_withIndex

시가     종가     고가    저가  rank_end
2016-03-01  10000  11000  13000  3000       4.0
2016-03-02  10300  12000  15000  5000       3.0
2016-03-03   9900  10000  15000  5000       5.0
2016-03-04  10500  13000  18000  8000       1.0
2016-03-5   11000  12000  17000  7000       3.0

### 참고)  중복 색인 : 있으면 있는대로, 없으면 없는대로!!!

In [18]:
# 중복 색인용 데이터 생성 - 앞의 부분 참조
# 중복 색인
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], 
                                        index = ["03-01","03-01","03-02","03-02","03-03"])
stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]

stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)

stock_price_DF_withIndex

시가     종가     고가    저가
03-01  10000  11000  13000  3000
03-01  10300  12000  15000  5000
03-02   9900  10000  15000  5000
03-02  10500  13000  18000  8000
03-03  11000  12000  17000  7000

#### 인덱스가 유일한지 확인해보자

In [20]:
stock_price_DF_withIndex.index.is_unique

False

In [19]:
# 유일한지 확인
stock_price_DF_withIndex.index.is_unique

False

#### 인덱스가 중복된 것의 자료를 조회해보자! : 인덱스 중복인 인덱스를 활용해 조회 vs 인덱스 유일한 것 활용해 조회 비교!!!!---> Type 또한 다름!!!! ---> 병합하거나 뭔가 조작을 할 때 Type Mismatch 의 위험이 있으니 주의 필요!!!!

In [21]:
stock_price_DF_withIndex.loc["03-01"]
# stock_price_DF_withIndex.ix["03-01"]

시가     종가     고가    저가
03-01  10000  11000  13000  3000
03-01  10300  12000  15000  5000

In [70]:
type(stock_price_DF_withIndex.loc["03-01"])
# type(stock_price_DF_withIndex.ix["03-01"])

pandas.core.frame.DataFrame

In [71]:
stock_price_DF_withIndex.loc["03-03"]
# stock_price_DF_withIndex.ix["03-03"]

시가    11000
종가    12000
고가    17000
저가     7000
Name: 03-03, dtype: int64

In [72]:
type(stock_price_DF_withIndex.loc["03-03"])
# type(stock_price_DF_withIndex.ix["03-03"])

pandas.core.series.Series

####  참고 : Skip) 뭔가 + 하면서 빠진 것들이 생기면 NA가 되고, 이는 앞의 fill_value 등으로 처리 가능함!


In [73]:
# fill_value
df1 = pd.DataFrame(np.arange(12).reshape((3,4)), columns=list("abcd"))
df2 = pd.DataFrame(np.arange(20).reshape((4,5)), columns=list("abcde"))
print(df1)
print(df2)

   a  b   c   d
0  0  1   2   3
1  4  5   6   7
2  8  9  10  11
    a   b   c   d   e
0   0   1   2   3   4
1   5   6   7   8   9
2  10  11  12  13  14
3  15  16  17  18  19


In [74]:
df1 + df2

a     b     c     d   e
0   0.0   2.0   4.0   6.0 NaN
1   9.0  11.0  13.0  15.0 NaN
2  18.0  20.0  22.0  24.0 NaN
3   NaN   NaN   NaN   NaN NaN

In [75]:
df1.add(df2, fill_value=100)

a      b      c      d      e
0    0.0    2.0    4.0    6.0  104.0
1    9.0   11.0   13.0   15.0  109.0
2   18.0   20.0   22.0   24.0  114.0
3  115.0  116.0  117.0  118.0  119.0

## 중요!!!!! 일괄 적용 부분!!!!

### 실습 데이터 만들기

In [23]:
# apply / applymap 등과 같은 일괄 적용하는 부분!!!!!!!! 
#  실습용 데이터 생성
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000],
                                        index = ["2016-03-03","2016-03-01","2016-03-02","2016-03-5","2016-03-04"])
stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [3000,5000,5000,8000,7000]
stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)
stock_price_DF_withIndex

시가     종가     고가    저가
2016-03-03  10000  11000  13000  3000
2016-03-01  10300  12000  15000  5000
2016-03-02   9900  10000  15000  5000
2016-03-5   10500  13000  18000  8000
2016-03-04  11000  12000  17000  7000

#### 요일별로 최고값과 최저값을 찾아서 그 차이를 계산해보자!!

In [26]:
stock_price_DF_withIndex.apply(lambda x : x.max() - x.min(), axis=1)

2016-03-03    10000
2016-03-01    10000
2016-03-02    10000
2016-03-5     10000
2016-03-04    10000
dtype: int64

In [77]:
print(stock_price_DF_withIndex.apply(lambda x:x.max()-x.min(), axis=1))


2016-03-03    10000
2016-03-01    10000
2016-03-02    10000
2016-03-5     10000
2016-03-04    10000
dtype: int64


#### 기준을 가로/ 세로 등으로 해보기!!! & 값들을 더하거나 곱하는 것들로 일괄 변경!!

In [78]:
# 여러 가지 적용한 결과 살펴보기 : 각기 하나씩 다 처리!!!!
print(stock_price_DF_withIndex.apply(lambda x:x.max()-x.min(), axis=0))
print("---------------------------------------------------------------")
print(stock_price_DF_withIndex.apply(lambda x:x.max()-x.min(), axis=1))
print("---------------------------------------------------------------")
print(stock_price_DF_withIndex.applymap(lambda x: x+1))
print("---------------------------------------------------------------")
print(stock_price_DF_withIndex.applymap(lambda x: x*100))

시가    1100
종가    3000
고가    5000
저가    5000
dtype: int64
---------------------------------------------------------------
2016-03-03    10000
2016-03-01    10000
2016-03-02    10000
2016-03-5     10000
2016-03-04    10000
dtype: int64
---------------------------------------------------------------
               시가     종가     고가    저가
2016-03-03  10001  11001  13001  3001
2016-03-01  10301  12001  15001  5001
2016-03-02   9901  10001  15001  5001
2016-03-5   10501  13001  18001  8001
2016-03-04  11001  12001  17001  7001
---------------------------------------------------------------
                 시가       종가       고가      저가
2016-03-03  1000000  1100000  1300000  300000
2016-03-01  1030000  1200000  1500000  500000
2016-03-02   990000  1000000  1500000  500000
2016-03-5   1050000  1300000  1800000  800000
2016-03-04  1100000  1200000  1700000  700000


#### Q) 위와 같은 데이터를 가지고, stock_Price_DF_withIndex에서 “Gap” 라는 컬럼을 생성하고, 이 컬럼에서는 시가/종가/고가/저가 중에서 가장 큰 값고 가장 작은 값의 차이를 기록하는 컬럼을 생성하시오.

In [28]:
# sol)
# 데이터 생성)
stock_price_DF_withIndex = pd.DataFrame([10000, 10300, 9900, 10500, 11000], index = ["2016-03-03","2016-03-01","2016-03-02","2016-03-05","2016-03-04"])
stock_price_DF_withIndex["종가"] = [11000,12000,10000,13000,12000]
stock_price_DF_withIndex["고가"] = [13000,15000,15000,18000,17000]
stock_price_DF_withIndex["저가"] = [6500,5300,8700,1000,7500]
stock_price_DF_withIndex.rename(columns={0:"시가"}, inplace = True)
print(stock_price_DF_withIndex)
print("---------------------------------------------------------------")

# 일괄 작업)
stock_price_DF_withIndex["Gap"] = stock_price_DF_withIndex.apply(lambda x:x.max()-x.min(), axis=1)


print(stock_price_DF_withIndex)

               시가     종가     고가    저가
2016-03-03  10000  11000  13000  6500
2016-03-01  10300  12000  15000  5300
2016-03-02   9900  10000  15000  8700
2016-03-05  10500  13000  18000  1000
2016-03-04  11000  12000  17000  7500
---------------------------------------------------------------
               시가     종가     고가    저가    Gap
2016-03-03  10000  11000  13000  6500   6500
2016-03-01  10300  12000  15000  5300   9700
2016-03-02   9900  10000  15000  8700   6300
2016-03-05  10500  13000  18000  1000  17000
2016-03-04  11000  12000  17000  7500   9500


In [80]:
stock_price_DF_withIndex

시가     종가     고가    저가    Gap
2016-03-03  10000  11000  13000  6500   6500
2016-03-01  10300  12000  15000  5300   9700
2016-03-02   9900  10000  15000  8700   6300
2016-03-05  10500  13000  18000  1000  17000
2016-03-04  11000  12000  17000  7500   9500

In [120]:
len(stock_price_DF_withIndex)

5

In [132]:
for i in stock_price_DF_withIndex.index:
    print(stock_price_DF_withIndex.loc[i,"시가":"저가"])

시가    10000
종가    11000
고가    13000
저가     6500
Name: 2016-03-03, dtype: int64
시가    10300
종가    12000
고가    15000
저가     5300
Name: 2016-03-01, dtype: int64
시가     9900
종가    10000
고가    15000
저가     8700
Name: 2016-03-02, dtype: int64
시가    10500
종가    13000
고가    18000
저가     1000
Name: 2016-03-05, dtype: int64
시가    11000
종가    12000
고가    17000
저가     7500
Name: 2016-03-04, dtype: int64


In [133]:
for i in range(0, len(stock_price_DF_withIndex)):
    print(stock_price_DF_withIndex.iloc[i, 0:4])

시가    10000
종가    11000
고가    13000
저가     6500
Name: 2016-03-03, dtype: int64
시가    10300
종가    12000
고가    15000
저가     5300
Name: 2016-03-01, dtype: int64
시가     9900
종가    10000
고가    15000
저가     8700
Name: 2016-03-02, dtype: int64
시가    10500
종가    13000
고가    18000
저가     1000
Name: 2016-03-05, dtype: int64
시가    11000
종가    12000
고가    17000
저가     7500
Name: 2016-03-04, dtype: int64


#### Q)이 이렇게 1~2줄로 완성할 것을 apply와 lambda를 사용하지 않고, for를 사용해서 코딩을 한다고 하면??직접 해보고 값을 비교할 것!!!


In [81]:
# 사전 확인 사항 : 적용할 대상 생각하기!!!
#stock_price_DF_withIndex.ix[1, 0:4]
stock_price_DF_withIndex.iloc[1, 0:4]

시가    10300
종가    12000
고가    15000
저가     5300
Name: 2016-03-01, dtype: int64

In [29]:
stock_price_DF_withIndex

시가     종가     고가    저가    Gap
2016-03-03  10000  11000  13000  6500   6500
2016-03-01  10300  12000  15000  5300   9700
2016-03-02   9900  10000  15000  8700   6300
2016-03-05  10500  13000  18000  1000  17000
2016-03-04  11000  12000  17000  7500   9500

In [82]:
stock_price_DF_withIndex["gap_forloop"] = 0
for i in range(0, len(stock_price_DF_withIndex)):
    i번째 줄에서 앞의 4줄의 값(i번째 줄의 시가, 종가, 고가, 저가) : iloc [i,~~]
    위의 4개 값 중에서 최고값(max) : .max
    위의 4개 값 중에서 최소값(min) : .min
    4개 값의 max - min : temp
    업데이트 할 위치 : 가로-i번째, 세로줄 - gap_forloop, 넣을 값 temp

SyntaxError: invalid syntax (<ipython-input-82-a8c063d8072d>, line 3)

* 업데이트 사항을 반영하여 그 전에 확인할 사항들!!!!

In [84]:
stock_price_DF_withIndex["Gap_forloop_test"] = 0
labels = ["Gap_forloop_test"]

# 참고) 가로 인덱스인 경우
# index_ints = [stock_price_DF_withIndex.index.get_loc(label) for label in labels]
# 우리는 세로의 컬럼 이름이니...
index_ints = [stock_price_DF_withIndex.columns.get_loc(label) for label in labels]
index_ints

[6]

In [3]:
# updated 방식 : 2018.08
# ---> 최근 업데이트를 반영하여 이렇게 수정하는 방식을 추천


# 위의 apply와 lambda를 적용하면, 1줄에 작성할 것을 이렇게 여러 줄로 해야함!!!!!
stock_price_DF_withIndex["Gap_forloop"] = 0
labels = ["Gap_forloop"]
index_ints = [stock_price_DF_withIndex.columns.get_loc(label) for label in labels]

for i in range(0, len(stock_price_DF_withIndex)):
    tempValue = stock_price_DF_withIndex.iloc[i,0:4]
    tempMax = tempValue.max()
    tempMin = tempValue.min()
    stock_price_DF_withIndex.iloc[i,index_ints[0]] = tempMax - tempMin
    
stock_price_DF_withIndex

시가     종가     고가    저가    Gap  Gap_forloop
2016-03-03  10000  11000  13000  6500   6500         6500
2016-03-01  10300  12000  15000  5300   9700         9700
2016-03-02   9900  10000  15000  8700   6300         6300
2016-03-05  10500  13000  18000  1000  17000        17000
2016-03-04  11000  12000  17000  7500   9500         9500

#### 참고)  주의!!!!!)위의 것은  ix 이용한 방법이나, ix 보다는 set_value 가 효율적이다. 그러나 이렇게 변경하기 위해서는 위의 코드도 일부 변경을 해야함!!! ---> 지금은 update가 되어서 ix는 사용하지 못하고,  set_vlaue도 사용불가능!!!! 

#### Q) 위의 데이터에서 “시가”, “종가”,”고가”,”저가”의 3월1일부터 5일까지의 평균을 구하여라!!!

* Hint) x를 기준으로 생각하면, 1개가 아니라

In [88]:
# 3월 1~ 5일에 걸치 시가의 평균을 구하라
# M1) numpy mean 이용
stock_price_DF_withIndex[["시가","종가","고가","저가"]].apply(lambda x : np.mean(x), axis=0)

시가    10340.0
종가    11600.0
고가    15600.0
저가     5800.0
dtype: float64

In [89]:
#M2) pandas Series 기본 mean 이용
stock_price_DF_withIndex[["시가","종가","고가","저가"]].apply(lambda x : x.mean(), axis=0)

시가    10340.0
종가    11600.0
고가    15600.0
저가     5800.0
dtype: float64

#### Q) 위의 데이터에서 일자별 시가/종가/고가/저가의 평균을 구하시오]

In [33]:
stock_price_DF_withIndex[["시가","종가","고가","저가"]].apply(lambda x : np.mean(x), axis=0)

시가    10340.0
종가    11600.0
고가    15600.0
저가     5800.0
dtype: float64

In [5]:
# 어떻게 해결할지 실마리!!!
#stock_price_DF_withIndex.ix[0,0:4].mean()
stock_price_DF_withIndex.iloc[0,0:4].mean()

10125.0

In [6]:
# M1)
stock_price_DF_withIndex[["시가","종가","고가","저가"]].apply(lambda x : np.mean(x), axis=1)

2016-03-03    10125.0
2016-03-01    10650.0
2016-03-02    10900.0
2016-03-05    10625.0
2016-03-04    11875.0
dtype: float64

In [7]:
# M2)
stock_price_DF_withIndex[["시가","종가","고가","저가"]].apply(lambda x : x.mean(), axis=1)

2016-03-03    10125.0
2016-03-01    10650.0
2016-03-02    10900.0
2016-03-05    10625.0
2016-03-04    11875.0
dtype: float64

## 간단한 기술 통계 : 자세한 사항은 google or manual

In [93]:
#  간단한 통계관련 부분
data = {'name': ['Jason', 'Molly', 'Tina', 'Jake', 'Amy'],
        'age': [42, 52, 36, 24, 73],
        'preTestScore': [4, 24, 31, 2, 3],
        'postTestScore': [25, 94, 57, 62, 70]}
df = pd.DataFrame(data, columns = ['name', 'age', 'preTestScore', 'postTestScore'])
df

name  age  preTestScore  postTestScore
0  Jason   42             4             25
1  Molly   52            24             94
2   Tina   36            31             57
3   Jake   24             2             62
4    Amy   73             3             70

In [94]:
#  간단한 통계관련 부분 - 01) Sum/mean/cumsum ...[p.195]
print (df['age'].sum())
print (df['preTestScore'].mean())
print (df['preTestScore'].cumsum())
print (df['preTestScore'].describe())


print (df['preTestScore'].var())
print (df['preTestScore'].std())
print (df['preTestScore'].skew())

227
12.8
0     4
1    28
2    59
3    61
4    64
Name: preTestScore, dtype: int64
count     5.000000
mean     12.800000
std      13.663821
min       2.000000
25%       3.000000
50%       4.000000
75%      24.000000
max      31.000000
Name: preTestScore, dtype: float64
186.7
13.663820841916802
0.7433452457326751


In [95]:
#  간단한 통계관련 부분 - 02) 상관관계/공분산
print (df.corr())
print (df.cov())

                    age  preTestScore  postTestScore
age            1.000000     -0.105651       0.328852
preTestScore  -0.105651      1.000000       0.378039
postTestScore  0.328852      0.378039       1.000000
                  age  preTestScore  postTestScore
age            340.80        -26.65         151.20
preTestScore   -26.65        186.70         128.65
postTestScore  151.20        128.65         620.30


* Unique 활용해보기

In [96]:
### Data Cleaning - unique 활용
print (stock_price_DF_withIndex)
print ("----------------------------")
print (stock_price_DF_withIndex["종가"])
print ("----------------------------")
print (stock_price_DF_withIndex["종가"].unique())
print ("----------------------------")
print (len(stock_price_DF_withIndex["종가"].unique()))

                 시가       종가       고가      저가      Gap  Gap_forloop  \
2016-03-03  10000.0  11000.0  13000.0  6500.0   6500.0         6500   
2016-03-01  10300.0  12000.0  15000.0  5300.0   9700.0         9700   
2016-03-02   9900.0  10000.0  15000.0  8700.0   6300.0         6300   
2016-03-05  10500.0  13000.0  18000.0  1000.0  17000.0        17000   
2016-03-04  11000.0  12000.0  17000.0  7500.0   9500.0         9500   
0               NaN      NaN      NaN     NaN      NaN            0   
1               NaN      NaN      NaN     NaN      NaN            0   
2               NaN      NaN      NaN     NaN      NaN            0   
3               NaN      NaN      NaN     NaN      NaN            0   
4               NaN      NaN      NaN     NaN      NaN            0   

            Gap_forloop_test  
2016-03-03                 0  
2016-03-01                 0  
2016-03-02                 0  
2016-03-05                 0  
2016-03-04                 0  
0                          0  
1

## Data Cleaning

* 빵구난 값 처리! 앞에 내용에 추가되어서!!!

* 어느 줄에서 빵구가 있는지를 기준으로!

In [2]:
### Data Cleaning
from numpy import nan as NA
import pandas as pd
data = pd.DataFrame([[1,6.5,3],[1,NA,NA],[NA,NA,NA],[NA,6.5,3]])
cleanded = data.dropna()

In [3]:
data

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
2  NaN  NaN  NaN
3  NaN  6.5  3.0

In [99]:
cleanded

0    1    2
0  1.0  6.5  3.0

In [100]:
# axis = 0으로 가로줄 기준으로 하나라고 NA가 있으면 제거
data.dropna(axis=0)

0    1    2
0  1.0  6.5  3.0

In [101]:
# axis = 1 으로 세로줄 기준으로 하나라도 NA가 있으면 제거
data.dropna(axis=1)

Empty DataFrame
Columns: []
Index: [0, 1, 2, 3]

In [102]:
# axis=0으로 가로줄인데, 가로줄의 모든 칸에 없는 것만 제거
data.dropna(how="all", axis=0)

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
3  NaN  6.5  3.0

In [103]:
# axis=1으로 세로줄인데, 세로줄의 모든 칸에 없는 것만 제거
data.dropna(how="all", axis=1)

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
2  NaN  NaN  NaN
3  NaN  6.5  3.0

* 빵구난 값에서 몇 개가 빵구났는지 기준으로 처리!

In [104]:
from numpy import nan as NA
df = pd.DataFrame(np.random.randn(7,3))
# df.ix[:4,1] = NA
# df.ix[:2,2] = NA
df.iloc[:4,1] = NA
df.iloc[:2,2] = NA
df

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [105]:
# 기본적으로 axis=0으로 가로줄을 기준으로 하나라도 있으면 제거
df.dropna()

0         1         2
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [106]:
# 기본으로 axis=0의 가로줄이 기준이나, 가로줄 기준으로 1개 이상의 정상 값이 있으면 살려두기 #
df.dropna(thresh=1)

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [107]:
# 기본으로 axis=0의 가로줄이 기준이나, 가로줄 기준으로 2개 이상의 정상 값이 있으면 살려두기 #
df.dropna(thresh=2)

0         1         2
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [108]:
# 기본으로 axis=0의 가로줄이 기준이나, 가로줄 기준으로 3개 이상의 정상 값이 있으면 살려두기 #
df.dropna(thresh=3)

0         1         2
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [109]:
#  axis=1의 세로줄 기준으로, 세로줄 기준으로 2개 이상의 정상 값이 있으면 살려두기 #
df.dropna(thresh=2, axis=1)

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [110]:
# 기본으로 axis=0의 가로줄이 기준이나, 가로줄 기준으로 3개 이상의 정상 값이 있으면 살려두기 #
df.dropna(thresh=3, axis=1)

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [111]:
# 1개의 값으로 채우기
df.fillna(1000)

0            1            2
0  1.126171  1000.000000  1000.000000
1 -0.260682  1000.000000  1000.000000
2  0.374949  1000.000000    -0.110607
3 -0.101090  1000.000000    -0.139092
4  1.253785     0.262877    -0.949996
5 -1.717610    -0.836630     1.575196
6 -0.107907    -0.363938    -0.572438

* fillna 등으로 빵구난 값을 채우기는 하는데, 모든 컬럼에 동일한 값으로 채우는 것이 아니라 컬러멩 따라서 다른 값으로 채워야 한다면????

* 어디에 뭐?? ---> Dict

In [112]:
# 데이터 처리할 때에는 컬럼별로 각기 처리하는 값이 다른 경우가 많음 
# --> 이런 경우에 처리하는 방법

# 1컬럼에는 9999로 처리, 3컬럼에는 -9999로 처리#
# 단, 3 컬럼은 없으니 에러 및 경고 없이 처리가 됨에 유의!!
df.fillna({1:9999,3:-9999})

0            1         2
0  1.126171  9999.000000       NaN
1 -0.260682  9999.000000       NaN
2  0.374949  9999.000000 -0.110607
3 -0.101090  9999.000000 -0.139092
4  1.253785     0.262877 -0.949996
5 -1.717610    -0.836630  1.575196
6 -0.107907    -0.363938 -0.572438

In [113]:
df.fillna(method="ffill")

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [114]:
# 뒤의 갚을 미러링해서 채우기
df.fillna(method="bfill")

0         1         2
0  1.126171  0.262877 -0.110607
1 -0.260682  0.262877 -0.110607
2  0.374949  0.262877 -0.110607
3 -0.101090  0.262877 -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [115]:
#보간법
df.interpolate()

0         1         2
0  1.126171       NaN       NaN
1 -0.260682       NaN       NaN
2  0.374949       NaN -0.110607
3 -0.101090       NaN -0.139092
4  1.253785  0.262877 -0.949996
5 -1.717610 -0.836630  1.575196
6 -0.107907 -0.363938 -0.572438

In [116]:
from numpy import nan as NA
df = pd.DataFrame(np.random.randn(7,3))
df.ix[1,1] = NA
df.ix[3,1] = NA
df.ix[2,2] = NA
df.ix[5,2] = NA

df

/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:4: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  after removing the cwd from sys.path.
/Users/minkyukwon/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:5: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/sta

0         1         2
0 -0.015567  1.583492  0.382958
1  1.535146       NaN -1.694167
2 -0.878220 -0.110494       NaN
3 -0.356849       NaN -0.891020
4 -0.619390 -1.581008  0.495153
5 -0.707430 -0.098723       NaN
6  0.422212  0.413909 -0.360318

In [117]:
df.interpolate()

0         1         2
0 -0.015567  1.583492  0.382958
1  1.535146  0.736499 -1.694167
2 -0.878220 -0.110494 -1.292593
3 -0.356849 -0.845751 -0.891020
4 -0.619390 -1.581008  0.495153
5 -0.707430 -0.098723  0.067418
6  0.422212  0.413909 -0.360318

In [ ]:
########################################################

In [ ]:
path = r"c:\data\tt.csv"